jax/_src/lax/lax.py

```diff
@@ -102,27 +102,36 @@ def _check_static_shape(shape: Shape):
   else:
     map(_check_static_shape, shapes)
 
-def _try_broadcast_shapes(
-    shapes: Sequence[tuple[int, ...]]) -> tuple[int, ...] | None:
-  if len(shapes) == 1: return shapes[0]
+def _try_broadcast_shapes(*shapes: tuple[int, ...],
+                          name: str,
+                          allow_rank_promotion: bool) -> tuple[int, ...]:
+  """
+  Attempt to broadcast shapes, raising a TypeError if broadcasting fails.
+  """
+  if not shapes:
+    raise TypeError(f"{name}: At least one shape required")
   ranks = {len(shape) for shape in shapes}
-  if len(ranks) > 1: return None  # must have consistent rank
-  rank = ranks.pop()
-  if not rank: return ()  # scalar case
+  if allow_rank_promotion:
+    ndim = _max(ranks)
+    shapes = tuple((*((1,) * (ndim - len(shape))), *shape) for shape in shapes)
```

Comment:
You can inline this below on line 177 and remove `allow_rank_promotion`? WDYT?

Reply:
I say this so that shape and sharding rules are similar to each other.

Reply:
Done, thanks!